srcbmi/mf6bmi.f90

```diff
@@ -593,6 +593,51 @@ function get_value_string(c_var_address, c_arr_ptr) result(bmi_status) &
 
   end function get_value_string
 
+  !> @brief Get a pointer to an array
+  !!
+  !! The array is located at @p c_var_address. There is no copying of data involved.
+  !! Multi-dimensional arrays are supported and the get_var_rank() function
+  !! can be used to get the variable's dimensionality, and get_var_shape() for
+  !! its shape.
+  !<
+  function get_value_ptr(c_var_address, c_arr_ptr) result(bmi_status) &
+    bind(C, name="get_value_ptr")
+    !DIR$ ATTRIBUTES DLLEXPORT :: get_value_ptr
+    ! -- modules
+    use ConstantsModule, only: LENMEMTYPE
+    ! -- dummy variables
+    character(kind=c_char), intent(in) :: c_var_address(*) !< memory address string of the variable
+    type(c_ptr), intent(inout) :: c_arr_ptr !< pointer to the array
+    integer(kind=c_int) :: bmi_status !< BMI status code
+    ! -- local variables
+    character(len=LENMEMPATH) :: mem_path
+    character(len=LENMEMTYPE) :: mem_type
+    character(len=LENVARNAME) :: var_name
+    logical(LGP) :: valid
+
+    bmi_status = BMI_SUCCESS
+
+    call split_address(c_var_address, mem_path, var_name, valid)
+    if (.not. valid) then
+      bmi_status = BMI_FAILURE
+      return
+    end if
+
+    call get_mem_type(var_name, mem_path, mem_type)
+
+    if (index(mem_type, "DOUBLE") /= 0) then
+      bmi_status = get_value_ptr_double(c_var_address, c_arr_ptr)
+    else if (index(mem_type, "INTEGER") /= 0) then
+      bmi_status = get_value_ptr_int(c_var_address, c_arr_ptr)
+    else
+      write (bmi_last_error, fmt_unsupported_rank) trim(var_name)
```

Comment:
This should be unsupported type, not rank

Reply:
> @mjr-deltares should we merge and deal with the deprecation warning after this is in?

after fixing the error message above, we can do that. But I realized while looking at it again that we also have the get_value_* functions for each of the types, and then also the set_value_* counter parts. We should then probably treat those similarly.